File: src/models/currency.py
import yfinance as yf

class Currency:
    def __init__(self) -> None:
        self.buffer = []
        
    def convert(self, value : float, from_currency : str, to_currency : str, exchange_rate_date : str) -> float:
        buffered = next((p for p in self.buffer if p[0] == from_currency and p[1] == to_currency and p[2] == exchange_rate_date), None)
        if buffered is not None:
            return value * buffered[3]
        src = from_currency
        dst = to_currency
        ticker = f'{src}{dst}=X'

        # Fetch the data
        data = yf.Ticker(ticker)

        # Get historical market data
        hist = data.history(start=exchange_rate_date)
        exchange_rate = hist.iloc[0]["Close"]
        self.buffer.append((src, dst, exchange_rate_date, exchange_rate))
        # print(f"Exchange rate for {src} to {dst} on {exchange_rate_date}: {exchange_rate}")
        return value * hist.iloc[0]["Close"]
    
    
if __name__ == "__main__":
    c = Currency()
    print(c.convert(100, 'USD', 'EUR', '2023-01-06'))

File: js/models/currency.js
import yf from "yahoo-finance2";

class Currency {
    constructor() {
        this.buffer = [];
    }

    async convert(value, fromCurrency, toCurrency, exchangeRateDate) {
        const buffered = this.buffer.find(p => p[0] === fromCurrency && p[1] === toCurrency && p[2] === exchangeRateDate);
        if (buffered) {
            return value * buffered[3];
        }

        const ticker = `${fromCurrency}${toCurrency}=X`;
        
        try {
            // Get next day of exchangeRateDate
            let date = new Date(exchangeRateDate);
            date.setDate(date.getDate() + 1);
            let exchangeRateDateNextDay = date.toISOString().split('T')[0];
            let data = await yf.chart(ticker, { period1: exchangeRateDate, period2: exchangeRateDateNextDay });
            if (data.length === 0) throw new Error("No exchange rate data found");
            
            let exchangeRate = data['quotes'][0].close;
            this.buffer.push([fromCurrency, toCurrency, exchangeRateDate, exchangeRate]);
            return value * exchangeRate;
        } catch (error) {
            console.error(`Error fetching exchange rate for ${fromCurrency} to ${toCurrency} on ${exchangeRateDate}:`, error);
            return null;
        }
    }
}

let currency = new Currency();
let value = await currency.convert(100, 'USD', 'EUR', '2023-01-06');
console.log(value);


export { Currency };